lib/ratfunul.gi

```diff
@@ -1069,18 +1069,72 @@ end);
 RedispatchOnCondition(Derivative,true,
   [IsPolynomial],[IsLaurentPolynomial],0);
 
-InstallOtherMethod(Derivative,"uratfun,ind",true,
+InstallOtherMethod(Derivative,"uratfun",true,
   [IsUnivariateRationalFunction],0,
 function(ratfun)
-local num,den;
+local num,den,R,cf,pow,dr,a;
+  # try to be good for the case of iterated derivatives by using if the
+  # denominator is a power
   num:=NumeratorOfRationalFunction(ratfun);
   den:=DenominatorOfRationalFunction(ratfun);
-  return (Derivative(num)*den-num*Derivative(den))/(den^2);
+  R:=CoefficientsRing(DefaultRing([den]));
+  cf:=Collected(Factors(den));
+  pow:=Gcd(List(cf,x->x[2]));
+  if pow>1 then
+    dr:=Product(List(cf,x->x[1]^(x[2]/pow))); # root
+  else
+    dr:=den;
+  fi;
+  #cf:=(Derivative(num)*dr-num*pow*Derivative(dr))/dr^(pow+1);
+  num:=Derivative(num)*dr-num*pow*Derivative(dr);
+  den:=dr^(pow+1);
+  if IsOne(Gcd(num,dr)) then
+    # avoid cancellation
+    num:=CoefficientsOfUnivariateLaurentPolynomial(num);
+    den:=CoefficientsOfUnivariateLaurentPolynomial(den);
+    cf:=UnivariateRationalFunctionByExtRepNC(FamilyObj(ratfun),
+      num[1],den[1],num[2]-den[2],
+        IndeterminateNumberOfUnivariateRationalFunction(ratfun));
+  else
+    cf:=num/den; # cancellation
+  fi;
+
+  # try to maintain factorization of denominator
+  den:=DenominatorOfRationalFunction(cf);
+  a:=den;
+  pow:=0;
+  while IsZero(a mod dr) do
+    pow:=pow+1;
+    a:=a/dr;
+  od;
+  if IsOne(a) then
+    StoreFactorsPol(R,den,ListWithIdenticalEntries(pow,dr));
+  fi;
+  return cf;
+  #return (Derivative(num)*den-num*Derivative(den))/(den^2);
 end);
 
 RedispatchOnCondition(Derivative,true,
   [IsPolynomial],[IsUnivariateRationalFunction],0);
 
+
+InstallGlobalFunction(TaylorSeriesRationalFunction,function(f,at,deg)
+local t,i,x;
+  if not IsUnivariateRationalFunction(f) then
+    Error("function is not univariate");
+  fi;
+  x:=X(LeftActingDomain(DefaultRing(f^0)),
+    IndeterminateNumberOfUnivariateRationalFunction(f));
+  t:=Zero(f);
+  for i in [0..deg] do
+    if i>0 then 
+      f:=Derivative(f);
+    fi;
+    t:=t+Value(f,at)*(x-at)^i/Factorial(i);
+  od;
```

Comment:
Assuming we require `deg >= 0` (which is not clear to me from the description of the function), could also do this
```suggestion
  t:=Value(f,at)*x^0;
  for i in [1..deg] do
    f:=Derivative(f);
    t:=t+Value(f,at)*(x-at)^i/Factorial(i);
  od;
```

Reply:
I think this is a wash in terms of cost or code readability, so I left it out. (Added test for deg\ge0)